src/qililab/qprogram/experiment_executor.py

```diff
@@ -0,0 +1,258 @@
+        for element in block.elements:
+            if isinstance(element, (ForLoop, Loop)):
+                self._traverse_and_prepare(element)
+            # Handle ExecuteQProgram operations and traverse their loops
+            if isinstance(element, ExecuteQProgram):
+                self._traverse_qprogram(element.qprogram.body)
+
+    def _traverse_qprogram(self, block: Block):
+        """Traverses a QProgram to gather loop information."""
+        if isinstance(block, ForLoop):
+            loop_values = self._inclusive_range(block.start, block.stop, block.step)
+            loop_label = block.variable.label
+            self.loop_values[loop_label] = loop_values
+        elif isinstance(block, Loop):
+            loop_values = block.values
+            loop_label = block.variable.label
+            self.loop_values[loop_label] = loop_values
+
+        # Recursively handle nested blocks within the QProgram
+        for element in block.elements:
+            if isinstance(element, Block):
+                self._traverse_qprogram(element)
+
+    def _traverse_and_store(self, block: Block, progress: Progress):
+        """Traverse blocks, store generated Python functions, and return the stored operations."""
+        stored_operations = []
+
+        if isinstance(block, (Loop, ForLoop)):
+            # Handle loops
+            stored_operations.extend(self._handle_loop(block, progress))
+        else:
+            # Handle generic blocks
+            stored_operations.extend(self._process_elements(block.elements, progress))
+
+        return stored_operations
+
+    def _handle_loop(self, block: ForLoop | Loop, progress: Progress) -> list[Callable]:
+        """Common logic for handling ForLoop and Loop blocks."""
+        stored_operations = []
+
+        # Determine loop parameters based on the type of block
+        loop_values: np.ndarray
+        if isinstance(block, ForLoop):
+            loop_values = self._inclusive_range(block.start, block.stop, block.step)
+        elif isinstance(block, Loop):
+            loop_values = block.values  # Assuming `block.values` is a numpy array or similar iterable
+
+        loop_label = block.variable.label
+
+        # Create the progress bar for the loop
+        def create_progress_bar():
+            total_iterations = len(loop_values)
+            loop_task_id = progress.add_task(f"Looping over {loop_label}", total=total_iterations)
+            self.task_ids[block.uuid] = loop_task_id  # Store the task ID associated with this loop block
+
+            # Track the index for this loop
+            self.loop_indices[loop_label] = 0
+
+            return loop_task_id
+
+        stored_operations.append(create_progress_bar)
+
+        def advance_progress_bar(variable_value: int | float) -> None:
+            loop_task_id = self.task_ids[block.uuid]
+            progress.update(loop_task_id, description=f"Looping over {loop_label}: {variable_value}")
```

Comment:
Could we fix the number of decimals here? Fabio got OCD seeing the video, me a little bit less but also. :D